File: hdk/node/src/plugin/response.ts
import { HookResponse, HookUsage } from "../protocol/messages";

export interface Usage {
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
}

export class Response {
  model: string;
  content: string;
  usage?: Usage;

  constructor(protoResp?: HookResponse) {
    this.model = protoResp?.model || "";
    this.content = protoResp?.content || "";
    if (protoResp?.usage) {
      this.usage = {
        promptTokens: protoResp.usage.prompt_tokens,
        completionTokens: protoResp.usage.completion_tokens,
        totalTokens: protoResp.usage.total_tokens,
      };
    }
  }

  toProtocol(): HookResponse {
    const proto: HookResponse = {
      model: this.model,
      content: this.content,
    };
    if (this.usage) {
      proto.usage = {
        prompt_tokens: this.usage.promptTokens,
        completion_tokens: this.usage.completionTokens,
        total_tokens: this.usage.totalTokens,
      };
    }
    return proto;
  }
}

File: hdk/go/plugin/response.go
package plugin

import "github.com/bhakha-services/naagmani-plugins/hdk/go/protocol"

// Usage carries token usage statistics.
type Usage struct {
	PromptTokens     int `json:"prompt_tokens"`
	CompletionTokens int `json:"completion_tokens"`
	TotalTokens      int `json:"total_tokens"`
}

// Response represents an AI model completion response.
type Response struct {
	Model   string `json:"model"`
	Content string `json:"content"`
	Usage   *Usage `json:"usage,omitempty"`
}

func responseFromProtocol(resp *protocol.HookResponse) *Response {
	if resp == nil {
		return nil
	}
	r := &Response{
		Model:   resp.Model,
		Content: resp.Content,
	}
	if resp.Usage != nil {
		r.Usage = &Usage{
			PromptTokens:     resp.Usage.PromptTokens,
			CompletionTokens: resp.Usage.CompletionTokens,
			TotalTokens:      resp.Usage.TotalTokens,
		}
	}
	return r
}

func (r *Response) toProtocol() *protocol.HookResponse {
	if r == nil {
		return nil
	}
	proto := &protocol.HookResponse{
		Model:   r.Model,
		Content: r.Content,
	}
	if r.Usage != nil {
		proto.Usage = &protocol.HookUsage{
			PromptTokens:     r.Usage.PromptTokens,
			CompletionTokens: r.Usage.CompletionTokens,
			TotalTokens:      r.Usage.TotalTokens,
		}
	}
	return proto
}
